File: src/clawhawk/sessions.py
"""JSONL session parser, caching, and grouped loading."""

from __future__ import annotations

import glob
import json
import logging
import os
import re
import time
from dataclasses import dataclass
from pathlib import Path

from clawhawk.ide import load_ide_map
from clawhawk.models import Message, ProjectGroup, Session

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Model context limits
# ---------------------------------------------------------------------------

_MODEL_CONTEXT_LIMITS: dict[str, int] = {
    "claude-opus-4-6": 1_000_000,
    "claude-sonnet-4-6": 200_000,
    "claude-haiku-4-5-20251001": 200_000,
}

_DEFAULT_CONTEXT_LIMIT = 200_000


def get_model_context_limit(model: str) -> int:
    """Return the context window size for a given model name."""
    return _MODEL_CONTEXT_LIMITS.get(model, _DEFAULT_CONTEXT_LIMIT)


# ---------------------------------------------------------------------------
# Text helpers
# ---------------------------------------------------------------------------

_RE_COMMAND_NAME = re.compile(r"<command-name>\s*(.*?)\s*</command-name>")


def clean_command_text(s: str) -> str:
    """Extract slash command from XML-tagged user messages, or return as-is."""
    m = _RE_COMMAND_NAME.search(s)
    if m is None:
        return s
    cmd = m.group(1).strip()
    return cmd if cmd else s


def truncate(s: str, max_len: int) -> str:
    """Truncate string, collapsing newlines to spaces."""
    s = s.replace("\n", " ").strip()
    if len(s) <= max_len:
        return s
    if max_len <= 3:
        return s[:max_len]
    return s[: max_len - 3] + "..."


def _truncate_keep_newlines(s: str, max_len: int) -> str:
    """Truncate string, preserving newlines."""
    s = s.strip()
    if len(s) <= max_len:
        return s
    if max_len <= 3:
        return s[:max_len]
    return s[: max_len - 3] + "..."


def decode_project_path(dirname: str) -> str:
    """Decode a project directory name back into a filesystem path.

    The directory name encodes a path by replacing "/" with "-".
    e.g. "-Users-tuongaz-dev-foo" -> "/Users/tuongaz/dev/foo"
    """
    if not dirname:
        return ""
    return dirname.replace("-", "/")


# ---------------------------------------------------------------------------
# Content extraction
# ---------------------------------------------------------------------------


def extract_user_text(content: object) -> str:
    """Extract user-visible text from a message content field."""
    if content is None:
        return ""

    # User messages typically have string content.
    if isinstance(content, str):
        return clean_command_text(content)

    # Sometimes content is an array of parts.
    if isinstance(content, list):
        for part in content:
            if isinstance(part, dict):
                if part.get("type") == "text":
                    text = part.get("text")
                    if isinstance(text, str):
                        return clean_command_text(text)

    return ""


def _extract_tool_detail(tool_name: str, tool_input: object) -> str:
    """Extract a short detail string from a tool_use input dict."""
    if not isinstance(tool_input, dict):
        return ""

    if tool_name in ("Read", "Edit", "Write"):
        fp = tool_input.get("file_path")
        if isinstance(fp, str):
            return truncate(fp, 120)
    elif tool_name == "Bash":
        desc = tool_input.get("description")
        if isinstance(desc, str) and desc:
            return truncate(desc, 120)
        cmd = tool_input.get("command")
        if isinstance(cmd, str):
            return truncate(cmd, 120)
    elif tool_name in ("Grep", "Glob"):
        pat = tool_input.get("pattern")
        if isinstance(pat, str):
            return truncate(pat, 80)
    elif tool_name == "Agent":
        desc = tool_input.get("description")
        if isinstance(desc, str):
            return truncate(desc, 120)
    elif tool_name in ("WebSearch", "WebFetch"):
        q = tool_input.get("query")
        if isinstance(q, str):
            return truncate(q, 120)
        u = tool_input.get("url")
        if isinstance(u, str):
            return truncate(u, 120)

    return ""


def extract_action(content: object) -> str:
    """Extract the last action description from an assistant message's content."""
    if content is None:
        return ""

    if isinstance(content, str):
        return truncate(content, 80)

    if not isinstance(content, list):
        return ""

    last_action = ""
    for part in content:
        if not isinstance(part, dict):
            continue

        part_type = part.get("type", "")
        if part_type == "tool_use":
            name = part.get("name")
            if isinstance(name, str):
                detail = _extract_tool_detail(name, part.get("input"))
                if detail:
                    last_action = name + ": " + detail
                else:
                    last_action = name
        elif part_type == "text":
            text = part.get("text")
            if isinstance(text, str) and text:
                last_action = truncate(text, 80)

    return last_action


# ---------------------------------------------------------------------------
# Session parser
# ---------------------------------------------------------------------------


def parse_session(fpath: str) -> Session | None:
    """Parse a JSONL session file into a Session object.

    Returns None if the file cannot be parsed or contains no valid session data.
    """
    first_prompt = ""
    last_user_prompt = ""
    last_action = ""
    last_ts = ""
    waiting_for_input = False
    last_model = ""
    last_context_tokens = 0

    session_id = ""
    session_name = ""
    cwd = ""
    git_branch = ""
    version = ""

    try:
        with open(fpath, encoding="utf-8", errors="replace") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue

                try:
                    msg = Message.model_validate_json(line)
                except Exception:
                    continue

                # Capture session metadata from first valid message.
                if not session_id and msg.session_id:
                    session_id = msg.session_id
                if msg.cwd:
                    cwd = msg.cwd
                if msg.git_branch:
                    git_branch = msg.git_branch
                if msg.timestamp:
                    last_ts = msg.timestamp
                if msg.version:
                    version = msg.version

                # Extract session name from agent-name entries.
                if msg.type == "agent-name" and msg.agent_name:
                    session_name = msg.agent_name

                # Track whether session is waiting for user input.
                if msg.type == "user":
                    waiting_for_input = False
                elif msg.type == "assistant" and msg.message.stop_reason == "end_turn":
                    waiting_for_input = True
                elif msg.type == "assistant" and msg.message.stop_reason == "tool_use":
                    waiting_for_input = False

                # Extract user prompts.
                if msg.type == "user":
                    text = extract_user_text(msg.message.content)
                    if text:
                        if not first_prompt:
                            first_prompt = truncate(text, 120)
                        last_user_prompt = _truncate_keep_newlines(text, 200)

                # Extract last action from assistant messages.
                if msg.type == "assistant":
                    action = extract_action(msg.message.content)
                    if action:
                        last_action = action

                    # Track token usage from the latest assistant message.
                    if msg.message.model:
                        last_model = msg.message.model
                    u = msg.message.usage
                    context_total = (
                        u.input_tokens
                        + u.cache_creation_input_tokens
                        + u.cache_read_input_tokens
                    )
                    if context_total > 0:
                        last_context_tokens = context_total

    except OSError:
        logger.warning("Failed to read session file: %s", fpath)
        return None

    if not session_id:
        return None

    return Session(
        session_id=session_id,
        name=session_name,
        cwd=cwd,
        git_branch=git_branch,
        timestamp=last_ts,
        first_prompt=first_prompt,
        last_user_prompt=last_user_prompt,
        last_action=truncate(last_action, 160),
        waiting_for_input=waiting_for_input,
        model=last_model,
        context_tokens=last_context_tokens,
        max_context_tokens=get_model_context_limit(last_model),
        version=version,
    )


# ---------------------------------------------------------------------------
# File-level cache
# ---------------------------------------------------------------------------


@dataclass
class CachedSession:
    mod_time: float
    session: Session


_cache: dict[str, CachedSession] = {}


# ---------------------------------------------------------------------------
# Active session detection
# ---------------------------------------------------------------------------


@dataclass
class ActiveInfo:
    """Active session detection data."""

    session_ids: set[str]
    running_cwds: set[str]


def _load_active_info(home: str) -> ActiveInfo:
    """Read ~/.claude/sessions/*.json to find sessions whose process is still running.

    Returns both the set of session IDs listed in the registry AND the set of
    cwds for running processes.  The cwds are used as a fallback: when /clear
    is used, Claude Code creates a new JSONL (with a new session ID) but does
    NOT update the registry file, so the new session ID won't appear in
    ``session_ids``.  We detect this by checking whether a JSONL belongs to a
    cwd that still has a live Claude process.
    """
    info = ActiveInfo(session_ids=set(), running_cwds=set())
    sessions_dir = os.path.join(home, ".claude", "sessions")

    try:
        entries = os.listdir(sessions_dir)
    except OSError:
        return info

    for name in entries:
        if not name.endswith(".json"):
            continue
        fpath = os.path.join(sessions_dir, name)
        try:
            with open(fpath, encoding="utf-8") as f:
                data = json.load(f)
        except (OSError, json.JSONDecodeError):
            continue

        session_id = data.get("sessionId", "")
        pid = data.get("pid", 0)
        cwd = data.get("cwd", "")
        if not session_id or not pid:
            continue

        # Check if process is still running.
        try:
            os.kill(pid, 0)
            info.session_ids.add(session_id)
            if cwd:
                info.running_cwds.add(cwd)
        except (OSError, ProcessLookupError):
            pass

    return info


# ---------------------------------------------------------------------------
# Grouped session loading
# ---------------------------------------------------------------------------


def load_grouped_sessions(limit: int = 0) -> list[ProjectGroup]:
    """Load all JSONL session files, group by project, and return sorted results.

    Globs ``~/.claude/projects/*/*.jsonl``, skips ``subagents/`` directories,
    caches parsed sessions by file modification time, enriches with IDE and
    memory information, and sorts by most recent timestamp descending.
    """
    home = os.path.expanduser("~")

    ide_dir = os.path.join(home, ".claude", "ide")
    ide_map = load_ide_map(ide_dir)
    active_info = _load_active_info(home)
    now = time.time()

    pattern = os.path.join(home, ".claude", "projects", "*", "*.jsonl")
    files = glob.glob(pattern)

    # Group sessions by project directory name.
    project_map: dict[str, list[Session]] = {}

    # First pass: parse all sessions and track the most recently modified
    # JSONL per running cwd (handles /clear creating a new session ID that
    # isn't in the registry).
    cwd_newest: dict[str, tuple[float, str]] = {}  # cwd -> (mod_time, session_id)
    parsed: list[tuple[str, Session]] = []  # (fpath, session)

    for fpath in files:
        if os.sep + "subagents" + os.sep in fpath:
            continue

        try:
            stat = os.stat(fpath)
        except OSError:
            continue

        mod_time = stat.st_mtime

        cached = _cache.get(fpath)
        if cached is not None and cached.mod_time == mod_time:
            sess = cached.session.model_copy()
        else:
            result = parse_session(fpath)
            if result is None:
                continue
            sess = result
            _cache[fpath] = CachedSession(mod_time=mod_time, session=sess.model_copy())

        if sess.cwd in active_info.running_cwds:
            prev = cwd_newest.get(sess.cwd)
            if prev is None or mod_time > prev[0]:
                cwd_newest[sess.cwd] = (mod_time, sess.session_id)

        parsed.append((fpath, sess))

    # Session IDs that are the newest file for a running cwd.
    cwd_active_ids: set[str] = {sid for _, sid in cwd_newest.values()}

    # Second pass: assign active status, project name, IDE client.
    for fpath, sess in parsed:
        is_registered = sess.session_id in active_info.session_ids
        is_newest_in_active_cwd = sess.session_id in cwd_active_ids
        sess.is_active = is_registered or is_newest_in_active_cwd

        dir_name = os.path.basename(os.path.dirname(fpath))
        sess.project_name = decode_project_path(dir_name)

        for folder, client in ide_map.items():
            if sess.cwd == folder or sess.cwd.startswith(folder + os.sep):
                sess.client = client
                break

        project_map.setdefault(dir_name, []).append(sess)

    # Build project groups.
    groups: list[ProjectGroup] = []

    for dir_name, sessions in project_map.items():
        # Sort sessions by timestamp descending.
        sessions.sort(key=lambda s: s.timestamp, reverse=True)

        if limit > 0 and len(sessions) > limit:
            sessions = sessions[:limit]

        # Check if this project has a memory directory with .md files.
        mem_dir = os.path.join(home, ".claude", "projects", dir_name, "memory")
        has_memory = False
        try:
            for entry in os.listdir(mem_dir):
                if entry.endswith(".md") and not os.path.isdir(
                    os.path.join(mem_dir, entry)
                ):
                    has_memory = True
                    break
        except OSError:
            pass

        if has_memory:
            for s in sessions:
                s.uses_memory = True

        decoded_path = decode_project_path(dir_name)
        groups.append(
            ProjectGroup(
                project_name=decoded_path,
                path=decoded_path,
                sessions=sessions,
            )
        )

    # Sort projects by most recent session timestamp descending.
    groups.sort(
        key=lambda g: g.sessions[0].timestamp if g.sessions else "",
        reverse=True,
    )

    return groups

File: internal/sessions/sessions.go
package sessions

import (
	"bufio"
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

type cachedSession struct {
	modTime time.Time
	session Session
}

var (
	cacheMu sync.Mutex
	cache   = make(map[string]cachedSession)
)

var modelContextLimits = map[string]int{
	"claude-opus-4-6":           1_000_000,
	"claude-sonnet-4-6":         200_000,
	"claude-haiku-4-5-20251001": 200_000,
}

const defaultContextLimit = 200_000

func getModelContextLimit(model string) int {
	if limit, ok := modelContextLimits[model]; ok {
		return limit
	}
	return defaultContextLimit
}

// Session represents a parsed Claude Code session.
type Session struct {
	SessionID   string `json:"sessionId"`
	ProjectName string `json:"projectName"`
	CWD         string `json:"cwd"`
	GitBranch   string `json:"gitBranch"`
	Timestamp   string `json:"timestamp"`
	FirstPrompt    string `json:"firstPrompt"`
	LastUserPrompt string `json:"lastUserPrompt"`
	LastAction     string `json:"lastAction"`
	IsActive        bool   `json:"isActive"`
	WaitingForInput bool   `json:"waitingForInput"`
	UsesMemory      bool   `json:"usesMemory"`
	Version          string `json:"version"`
	ContextTokens    int    `json:"contextTokens"`
	MaxContextTokens int    `json:"maxContextTokens"`
	Model            string `json:"model"`
	Client           string `json:"client"`
}

// ProjectGroup groups sessions by project directory.
type ProjectGroup struct {
	ProjectName string    `json:"projectName"`
	Path        string    `json:"path"`
	Sessions    []Session `json:"sessions"`
}

type message struct {
	Type      string         `json:"type"`
	SessionID string         `json:"sessionId"`
	CWD       string         `json:"cwd"`
	GitBranch string         `json:"gitBranch"`
	Timestamp string         `json:"timestamp"`
	Message   messageContent `json:"message"`
	Version   string         `json:"version"`
}

type messageUsage struct {
	InputTokens              int `json:"input_tokens"`
	CacheCreationInputTokens int `json:"cache_creation_input_tokens"`
	CacheReadInputTokens     int `json:"cache_read_input_tokens"`
	OutputTokens             int `json:"output_tokens"`
}

type messageContent struct {
	Role       string       `json:"role"`
	Content    any          `json:"content"`
	StopReason string       `json:"stop_reason"`
	Model      string       `json:"model"`
	Usage      messageUsage `json:"usage"`
}

// LoadGroupedSessions reads all JSONL session files from ~/.claude/projects/,
// parses them, groups by project, and returns sorted results.
// If limit > 0, only that many sessions per project are returned.
func LoadGroupedSessions(limit int) ([]ProjectGroup, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, err
	}

	ideDir := filepath.Join(home, ".claude", "ide")
	ideMap := loadIDEMap(ideDir)

	pattern := filepath.Join(home, ".claude", "projects", "*", "*.jsonl")
	files, err := filepath.Glob(pattern)
	if err != nil {
		return nil, err
	}

	// Group sessions by project directory.
	projectMap := make(map[string][]Session)
	for _, f := range files {
		// Skip anything under a subagents/ directory.
		if strings.Contains(f, string(filepath.Separator)+"subagents"+string(filepath.Separator)) {
			continue
		}

		info, err := os.Stat(f)
		if err != nil {
			continue
		}

		cacheMu.Lock()
		cached, ok := cache[f]
		cacheMu.Unlock()

		var sess *Session
		if ok && cached.modTime.Equal(info.ModTime()) {
			s := cached.session
			sess = &s
		} else {
			sess, err = parseSession(f)
			if err != nil {
				continue
			}
			cacheMu.Lock()
			cache[f] = cachedSession{modTime: info.ModTime(), session: *sess}
			cacheMu.Unlock()
		}

		// Recompute isActive since it's time-dependent.
		if sess.Timestamp != "" {
			if t, err := time.Parse(time.RFC3339Nano, sess.Timestamp); err == nil {
				sess.IsActive = time.Since(t) < 5*time.Minute
			}
		}

		dirName := filepath.Base(filepath.Dir(f))
		sess.ProjectName = decodeProjectPath(dirName)
		if client, ok := ideMap[sess.CWD]; ok {
			sess.Client = client
		}
		projectMap[dirName] = append(projectMap[dirName], *sess)
	}

	// Build project groups.
	var groups []ProjectGroup
	for dirName, sessions := range projectMap {
		// Sort sessions by timestamp descending.
		sort.Slice(sessions, func(i, j int) bool {
			return sessions[i].Timestamp > sessions[j].Timestamp
		})

		if limit > 0 && len(sessions) > limit {
			sessions = sessions[:limit]
		}

		// Check if this project has a memory directory with files.
		memDir := filepath.Join(home, ".claude", "projects", dirName, "memory")
		hasMemory := false
		if entries, err := os.ReadDir(memDir); err == nil {
			for _, e := range entries {
				if !e.IsDir() && strings.HasSuffix(e.Name(), ".md") {
					hasMemory = true
					break
				}
			}
		}
		if hasMemory {
			for i := range sessions {
				sessions[i].UsesMemory = true
			}
		}

		decodedPath := decodeProjectPath(dirName)
		groups = append(groups, ProjectGroup{
			ProjectName: decodedPath,
			Path:        decodedPath,
			Sessions:    sessions,
		})
	}

	// Sort projects by most recent session timestamp descending.
	sort.Slice(groups, func(i, j int) bool {
		if len(groups[i].Sessions) == 0 {
			return false
		}
		if len(groups[j].Sessions) == 0 {
			return true
		}
		return groups[i].Sessions[0].Timestamp > groups[j].Sessions[0].Timestamp
	})

	return groups, nil
}

func parseSession(fpath string) (*Session, error) {
	f, err := os.Open(fpath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var (
		sess              Session
		firstPrompt       string
		lastUserPrompt    string
		lastAction        string
		lastTS            string
		waitingForInput   bool
		lastModel         string
		lastContextTokens int
	)

	scanner := bufio.NewScanner(f)
	// Increase buffer size for potentially large lines.
	scanner.Buffer(make([]byte, 0, 1024*1024), 1024*1024)

	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}

		var msg message
		if err := json.Unmarshal(line, &msg); err != nil {
			continue
		}

		// Capture session metadata from first valid message.
		if sess.SessionID == "" && msg.SessionID != "" {
			sess.SessionID = msg.SessionID
		}
		if msg.CWD != "" {
			sess.CWD = msg.CWD
		}
		if msg.GitBranch != "" {
			sess.GitBranch = msg.GitBranch
		}
		if msg.Timestamp != "" {
			lastTS = msg.Timestamp
		}
		if msg.Version != "" {
			sess.Version = msg.Version
		}

		// Track whether session is waiting for user input.
		if msg.Type == "user" {
			waitingForInput = false
		} else if msg.Type == "assistant" && msg.Message.StopReason == "end_turn" {
			waitingForInput = true
		} else if msg.Type == "assistant" && msg.Message.StopReason == "tool_use" {
			waitingForInput = false
		}

		// Extract user prompts.
		if msg.Type == "user" {
			text := extractUserText(msg.Message.Content)
			if text != "" {
				if firstPrompt == "" {
					firstPrompt = truncate(text, 120)
				}
				lastUserPrompt = truncateKeepNewlines(text, 200)
			}
		}

		// Extract last action from assistant messages.
		if msg.Type == "assistant" {
			action := extractAction(msg.Message.Content)
			if action != "" {
				lastAction = action
			}

			// Track token usage from the latest assistant message.
			if msg.Message.Model != "" {
				lastModel = msg.Message.Model
			}
			u := msg.Message.Usage
			contextTotal := u.InputTokens + u.CacheCreationInputTokens + u.CacheReadInputTokens
			if contextTotal > 0 {
				lastContextTokens = contextTotal
			}
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	if sess.SessionID == "" {
		return nil, os.ErrNotExist
	}

	sess.Timestamp = lastTS
	sess.FirstPrompt = firstPrompt
	sess.LastUserPrompt = lastUserPrompt
	sess.LastAction = truncate(lastAction, 160)
	sess.WaitingForInput = waitingForInput
	sess.Model = lastModel
	sess.ContextTokens = lastContextTokens
	sess.MaxContextTokens = getModelContextLimit(lastModel)

	// Determine if the session is active (last message within 5 minutes).
	if lastTS != "" {
		if t, err := time.Parse(time.RFC3339Nano, lastTS); err == nil {
			sess.IsActive = time.Since(t) < 5*time.Minute
		}
	}

	return &sess, nil
}

func extractUserText(content any) string {
	if content == nil {
		return ""
	}

	// User messages typically have string content.
	if s, ok := content.(string); ok {
		return s
	}

	// Sometimes content is an array of parts.
	if parts, ok := content.([]any); ok {
		for _, part := range parts {
			if m, ok := part.(map[string]any); ok {
				if t, ok := m["type"].(string); ok && t == "text" {
					if text, ok := m["text"].(string); ok {
						return text
					}
				}
			}
		}
	}

	return ""
}

func extractAction(content any) string {
	if content == nil {
		return ""
	}

	// If content is a string, return a snippet.
	if s, ok := content.(string); ok {
		return truncate(s, 80)
	}

	// Content is an array of parts; find the last tool_use or text.
	parts, ok := content.([]any)
	if !ok {
		return ""
	}

	var lastAction string
	for _, part := range parts {
		m, ok := part.(map[string]any)
		if !ok {
			continue
		}

		t, _ := m["type"].(string)
		switch t {
		case "tool_use":
			if name, ok := m["name"].(string); ok {
				detail := extractToolDetail(name, m["input"])
				if detail != "" {
					lastAction = name + ": " + detail
				} else {
					lastAction = name
				}
			}
		case "text":
			if text, ok := m["text"].(string); ok && text != "" {
				lastAction = truncate(text, 80)
			}
		}
	}

	return lastAction
}

func extractToolDetail(toolName string, input any) string {
	m, ok := input.(map[string]any)
	if !ok || m == nil {
		return ""
	}

	switch toolName {
	case "Read", "Edit", "Write":
		if fp, ok := m["file_path"].(string); ok {
			return truncate(fp, 120)
		}
	case "Bash":
		if desc, ok := m["description"].(string); ok && desc != "" {
			return truncate(desc, 120)
		}
		if cmd, ok := m["command"].(string); ok {
			return truncate(cmd, 120)
		}
	case "Grep":
		if pat, ok := m["pattern"].(string); ok {
			return truncate(pat, 80)
		}
	case "Glob":
		if pat, ok := m["pattern"].(string); ok {
			return truncate(pat, 80)
		}
	case "Agent":
		if desc, ok := m["description"].(string); ok {
			return truncate(desc, 120)
		}
	case "WebSearch", "WebFetch":
		if q, ok := m["query"].(string); ok {
			return truncate(q, 120)
		}
		if u, ok := m["url"].(string); ok {
			return truncate(u, 120)
		}
	}

	return ""
}

func decodeProjectPath(dirname string) string {
	if dirname == "" {
		return ""
	}

	// The directory name encodes a path by replacing "/" with "-".
	// e.g., "-Users-tuongaz-dev-foo" -> "/Users/tuongaz/dev/foo"
	// A leading "-" indicates an absolute path starting with "/".
	result := strings.ReplaceAll(dirname, "-", "/")
	return result
}

func truncate(s string, max int) string {
	s = strings.ReplaceAll(s, "\n", " ")
	s = strings.TrimSpace(s)
	if len(s) <= max {
		return s
	}
	if max <= 3 {
		return s[:max]
	}
	return s[:max-3] + "..."
}

func truncateKeepNewlines(s string, max int) string {
	s = strings.TrimSpace(s)
	if len(s) <= max {
		return s
	}
	if max <= 3 {
		return s[:max]
	}
	return s[:max-3] + "..."
}
